### PR TITLE
Prototype http server

### PR DESCRIPTION
**Pull request by fritzm**

### CODE REVIEW

File: http/AjaxEndpoint.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// Local headers
+#include "http/Response.hpp"
+
+namespace lsst {
+namespace qserv {
+namespace http {
+
+class Server;
+
+class AjaxEndpoint
```
Comment (jgates108) [inline http/AjaxEndpoint.hpp:19]: Brief description of AjaxEndpoint, Path, etc. would be nice.

File: http/Request.cc [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// Third-party headers
+#include "boost/regex.hpp"
+
+namespace lsst {
+namespace qserv {
+namespace http {
+
+Request::Request(std::shared_ptr<boost::asio::ip::tcp::socket> socket)
+:
+    content(&_requestbuf),
```
Comment (jgates108) [inline http/Request.cc:13]: Reason not to use curly brackets? Move content onto same line as ':'?

File: http/AjaxEndpoint.hpp (hunk 1 above)
Comment (fritzm) [inline http/AjaxEndpoint.hpp:19]: Yes, more documentation everywhere is a must todo

File: http/Request.cc (hunk 2 above)
Comment (fritzm) [inline http/Request.cc:13]: I'm not sure about the curly inits yet -- see modern effective c++ article for the pros/cons?  That said, if we pick a style for the team and apply consistently I'm okay with that either way.  Let's discuss at the meeting tomorrow and I'm happy to convert this code if the team concensus is pre-curly-init.

File: http/Response.cc [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+namespace lsst {
+namespace qserv {
+namespace http {
+
+Response::Response(
+    std::shared_ptr<boost::asio::ip::tcp::socket> socket,
+    DoneCallback const &doneCallback)
+:
+    _socket(socket),
+    _doneCallback(doneCallback)
```
Comment (jgates108) [inline http/Response.cc]: Is it possible for doneCallback to fall out of scope?

Comment (jgates108) [inline http/Response.cc:73]: nvm, I see it's copied.

**Comment by ktlim** [comment]
I'm a bit concerned. Nothing like this already exists? I don't really want to be responsible for maintaining HTTP response code tables, URL parsers, etc.

**Comment by fritzm** [comment]
@ktlim So, I -did- look around pretty hard, and everything that I found was unsatisfactory for one or more of the following reasons:
- Many didn't use boost::asio, so wouldn't interact naturally with the threading/dispatch model being used in the rest of the prototype code.
- Many others were large and lumbering, way too heavy-weight for this intended application (light-weight embedding in services for monitoring/debug/admin).
- The usual egg-salad of half-abandoned / half-implemented open source projects.
- Several I tried were buggy of such poor coding style that I would have felt a strong obligation to fork and fix anyway, and probably to an extent that changes wouldn't be accepted upstream in a timely fashion.
- Very suitable packages do exist for Node, Java, Python, and Go, but I could not find the right thing for C++.

In the end, it seemed equivalent or less work to just craft exactly what I needed.  It will match our coding standards, can use our logging infrastructure etc.  It will be easier to maintain (and from what I've seen out there I'd wager less buggy as well.)  And we're only talking about a few hundreds of lines of code here.  This is not intended to be a client-facing web server, so it doesn't need every feature under the sun.

I will reassure you that I do consciously try to pick my battles and avoid NIH syndrome.  Some of the time though, after doing the diligence, if you want the right thing you just have to go ahead and write it!

(ob smiley for lightness of tone: :smiley:)

**Comment by fritzm** [comment]
@jgates108 per hallway discussions, link below for some example code that uses this (sets up a web server with a few REST endpoints, some static content, and an AJAX endpoint):

https://github.com/lsst-dm/qrepl/blob/master/qreplmon/QReplMon.cc#L61-73

**Pull request closed without merging**